portal-web/docroot/html/common/themes/user_locale_options.jsp

```diff
@@ -47,14 +45,11 @@ String currentURL = PortalUtil.getCurrentURL(request);
 	</div>
 
 	<aui:script use="aui-base,liferay-store">
-		var ignoreUserLocaleOptionsNode = A.one('#ignoreUserLocaleOptions');
-
-		ignoreUserLocaleOptionsNode.on(
-			'click',
-			function() {
+		Liferay.once('toastClosed', function (event) {
```

Comment:
If we did  https://github.com/liferay-frontend/liferay-portal/pull/39/files#r443635470, then we would "simply" need to extract this to a `user_locale_options_on_close.jsp`... which will look completely off for a JSP, but will avoid having to link this JSP through a global `toastClosed` API.

Reply:
@brunobasto, @diegonvs, @wincent... thoughts?

Reply:
You mean [here](https://github.com/liferay/liferay-portal/blob/f380f73537a0445f153e7a055fa36438b312bf69/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L830) we would need to pass an additional attribute (`onCloseJspPath`) to the JSP? And probably add a new constant `PortalMessages.ON_CLOSE_JSP_PATH` to follow the pattern there? Feels dangerous. Those constants are exposed API and if we decide to take it out anytime it will be a major version semantic change.

I'd rather having a way to identify a toast and listen for events when it opens or closes. Something like 

```js
Liferay.openToast({
...
id / key: ''userLocaleToast'
...
})
```

But anyway we'd still need to pass some kind of value from https://github.com/liferay/liferay-portal/blob/f380f73537a0445f153e7a055fa36438b312bf69/portal-impl/src/com/liferay/portal/util/PortalImpl.java#L830 to the JSP. So maybe we really have to add another constant. Or, we could just set a request attribute maybe.

Reply:
> You mean here we would need to pass an additional message (onCloseJspPath) to the JSP? And probably add a new constant PortalMessages.ON_CLOSE_JSP_PATH to follow the pattern there? Feels dangerous. Those constants are exposed API and if we decide to take it out anytime it will be a major version semantic change.

Yeah... just trying to stay closer to what we have here... which is in fact just 2 instances. Only 2 places we use `PortalMessages.add` in the whole portal 😢

Reply:
Another though. What if we don't use the generic messages mechanism for this specific use case and make `user_locale_options.jsp` do its own call to `Liferay.openToast()` and in there it could add an `onClose` listener?

The tricky part would be to figure out how that JSP would be included on the page. I only found two usages of `PortalMessages.KEY_JSP_PATH` so maybe we can assume that whenever we need to include a jsp in PortalMessages, that jsp will be responsible for opening its own toast?

Reply:
> I'd rather having a way to identify a toast and listen for events when it opens or closes.

We kind of have that already. You can pass a `containerId` when you call `openToast`, and @diegonvs has already added in this PR a `Liferay.fire` call that could be extended to include that in the callback data. (It is already passing `renderData`, which includes a `portletId`, which itself may be identifying enough, depending how many toasts a given portlet may render; and if it's not identifying enough, you can pass arbitrary data in `renderData`). Unless I am miscounting, that is three different ways to identify an individual toast to a greater or lesser degree.

Reply:
> I'd rather having a way to identify a toast and listen for events when it opens or closes. Something like

We've done this in some other places... `openToast` would do a `Liferay.component` call registering some type of proxy... and in this case that would mean wrapping the `onClose` prop to run it and fire an additional event so you could do `Liferay.componetnReady('id').on('close'...)`. 

While this is works, it favours a programming style that we're trying to run away from, so I'd want to defer the need for such APIs as much as we can.

Reply:
> The tricky part would be to figure out how that JSP would be included on the page. I only found two usages of PortalMessages.KEY_JSP_PATH so maybe we can assume that whenever we need to include a jsp in PortalMessages, that jsp will be responsible for opening its own toast?

Yeah, this is where this is inevitably leading. The more troubling issue is that this will likely cause a breaking change.

One option would be migrate the logic to a `TopHeadDynamicInclude` so we can control when it gets inserted and as you say, do the `openToast` call itself. Being in `portal-web`, though, not sure how much we can do or where we could move this to.

Reply:
So, what if we just leverage the `jspPath` being passed and just include it like the the snippet I drafted above? With that the breaking change would be the expected behavior of passing `PortalMessages.KEY_JSP_PATH `. Before it was assumed that we'd open the toast for them. With that change they'd have to open the toast themselves. It is a breaking change, but it doesn't look like passing `PortalMessages.KEY_JSP_PATH` is very popular.

Reply:
So, this is only used from `top_messages.jsp` as:

```
<%
if (!user.isDefaultUser() && !locale.equals(user.getLocale())) {
	PortalUtil.addUserLocaleOptionsMessage(request);
}
%>
```

So... I'd say we can move this out to an OSGi module with a `TopHeadDynamicInclude` and leave everything else as is, without expanding or breaking the existing API... as limited as it is :/

Reply:
Sounds reasonable. The other included JSP I found was https://github.com/liferay/liferay-portal/blob/master/portal-web/docroot/html/common/themes/layout_friendly_url_redirect.jsp but it doesn't do any JS so we're good there.

Reply:
Ok, asking for feedback on where to place this... I'll send a followup once that's clear because right now I honestly couldn't tell...

@diegonvs, I'll send a followup with some of the changes here but keeping the current API as stable as possible and we can continue conversing there, ok!?

Reply:
Ok @jbalsas!